### PR TITLE
New query design integration tests

### PR DESCRIPTION
**Pull request by suresh-prakash**
## Description
* Integrated the new generic query design with the existing interface
* Addressed follow-up comments from the previous PR (#58)
* Added necessary integration tests, along with a few bug-fixes.

### Testing
Added integration tests

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

### CODE REVIEW

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java [hunk 1]
```diff
@@ -55,81 +50,211 @@
...
+ *         .offset(5)
+ *         .limit(10)
  *         .build();
  *  </code>
  */
-@Value
 @AllArgsConstructor(access = AccessLevel.PRIVATE)
-@Builder
-public class Query {
+public final class Query {
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:66]: We can change the following as discussed :
- Method names to something like addSelection, setFilter etc to explicitly convey that one adds to a list and the other can reset the previously set value. 
- We can remove filters() method to avoid all error checking to support this
- Intermediate Selection type class added may not be required 
- For consistency we can have addSelection(selection) and addSelection(List<selection>)

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java [hunk 2]
```diff
@@ -55,81 +50,241 @@
  *                 AggregateExpression.of(SUM, IdentifierExpression.of("col5")),
  *                 GTE,
  *                 ConstantExpression.of(100)))
- *         .sortingDefinition(
+ *         .sort(
```
Comment (varsha-abhinandan) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:53]: The comment section needs to be updated.

Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:53]: Good catch.

Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java]: Done.

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java (hunk 1 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:65]: Instead of 
`Intermediate Selection type class added may not be required`, exposed the definition classes outside as discussed.

The remaining 3 items are done.

**Review by varsha-abhinandan — DISMISSED** [review]

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java [hunk 3]
```diff
@@ -275,6 +275,16 @@ public BulkUpdateResult bulkOperationOnArrayValue(BulkArrayValueUpdateRequest re
     return Collections.emptyIterator();
   }
 
+  @Override
+  public Iterator<Document> find(org.hypertrace.core.documentstore.query.Query query) {
```
[the remark this answers is not in the document]
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java:279]: Sure, Ronak. Created here: https://github.com/hypertrace/document-store/issues/61

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+            SelectionSpec.of(IdentifierExpression.of("quantity")),
+            SelectionSpec.of(IdentifierExpression.of("date")));
+    Selection selection = Selection.builder().selectionSpecs(selectionSpecs).build();
+    Filter filter =
+        Filter.builder()
+            .expression(
+                RelationalExpression.of(
+                    IdentifierExpression.of("item"),
+                    NOT_IN,
+                    ConstantExpression.ofStrings(List.of("Soap", "Bottle"))))
```
Comment (kotharironak) [inline document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:117]: nit: for completeness that it matches more than one, we could have one more matching `shampoo in the list.

Comment (kotharironak) [reply document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:117]: Just saw that we have a test for this later.

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    Query query =
+        Query.builder()
+            .setSelection(selection)
+            .setFilter(filter)
+            .setSort(sort)
+            .setPagination(pagination)
+            .build();
+
+    Iterator<Document> resultDocs = collection.find(query);
+    assertDocsEqual(resultDocs, "mongo/filter_with_sorting_and_pagination_response.json");
```
Comment (kotharironak) [inline document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:162]: The doc id `4` with item `Shampoo` has quantity `20`, why is that on the list?
```
{
    "_id": 4,
    "item": "Shampoo",
    "price": 5,
    "quantity": 20,
    "date": "2014-04-04T11:21:39.736Z"
  }
```
`Mirror` is offset out, but `Shampoo` should have been there, right?

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+                            IN,
+                            ConstantExpression.ofStrings(List.of("Mirror", "Comb", "Shampoo"))))
+                    .operator(OR)
+                    .operand(
+                        RelationalExpression.of(
+                            IdentifierExpression.of("props.seller.address.pincode"),
+                            EQ,
+                            ConstantExpression.of(700007)))
+                    .build())
+            .addSort(IdentifierExpression.of("props.brand"), ASC)
```
Comment (kotharironak) [inline document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:190]: nit: `props.brand` and `props.seller.address.city` should have been also added in selection spec, so it's easy to see the order result for comparision.

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+                            IdentifierExpression.of("total"), GTE, ConstantExpression.of(25)))
+                    .operator(AND)
+                    .operand(
+                        RelationalExpression.of(
+                            IdentifierExpression.of("total"), LTE, ConstantExpression.of(99)))
+                    .build())
+            .setFilter(
+                RelationalExpression.of(
+                    IdentifierExpression.of("quantity"), NEQ, ConstantExpression.of(10)))
+            .setLimit(10)
```
Comment (kotharironak) [inline document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:247]: It seems we now two way to set
- `limit` and `offset`
- `pagination`

What is the difference? When should be used explicitly?
If, both are interchangeable, can we just have a single way?

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    Iterator<Document> resultDocs = collection.aggregate(query);
+    assertDocsEqual(resultDocs, "mongo/sum_response.json");
+  }
+
+  @Test
+  public void testAggregateWithNestedFields() throws IOException {
+    Query query =
+        Query.builder()
+            .addSelection(
+                AggregateExpression.of(FIRST, IdentifierExpression.of("item")), "first_item")
```
Comment (kotharironak) [inline document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:260]: How easy would this be to do in `Postgres`? It is group by, right?

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/AggregationOperator.java [hunk 9]
```diff
@@ -6,4 +6,5 @@
   SUM,
   MIN,
   MAX,
+  FIRST,
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/AggregationOperator.java:9]: will we be able to handle this in `Postgres`?

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoSelectingExpressionParser.java [hunk 10]
```diff
@@ -7,71 +7,68 @@
...
+    return selectionSpecs.stream()
+        .filter(not(SelectionSpec::isAggregation))
+        .map(exp -> exp.getExpression().parse(parser).toString())
+        .collect(Collectors.toList());
   }
 
-  public static BasicDBObject getProjectClause(final List<Selection> selections) {
-    if (selections.stream().anyMatch(Selection::allColumnsSelected)
-        || selections.stream().allMatch(Selection::isAggregation)) {
+  public static BasicDBObject getProjectClause(final Query query) {
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoSelectingExpressionParser.java:65]: nit : `getProjectClause` -> `getProjectionClause`

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java [hunk 11]
```diff
@@ -34,102 +28,260 @@
...
 @AllArgsConstructor(access = AccessLevel.PRIVATE)
-@Builder
-public class Query {
+public final class Query {
+  private final Selection selection; // Missing selection represents fetching all the columns
+  private final Filter filter;
 
-  @NotEmpty List<@NotNull Selection> selections;
+  private final Aggregation aggregation;
+  private final Filter aggregationFilter;
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:70]: Just curious: Any specific reason for having `filter` and `aggregation filter` separate?

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java (hunk 5 above)
Comment (suresh-prakash) [reply document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:162]: We sort all the documents based on the descending order of quantity (and then, by ascending order of item).

So, after sorting, we have the following list:
`[
  {
    "_id": 4,
    "item": "Shampoo",
    "price": 5,
    "quantity": 20,
    "date": "2014-04-04T11:21:39.736Z"
  },
  {
    "_id": 7,
    "item": "Comb",
    "price": 7.5,
    "quantity": 10,
    "date": "2015-09-10T08:43:00Z",
    "props": {
      "seller": {
        "name": "Go Go Plastics",
        "address": {
          "city": "Kolkata",
          "pincode": 700007
        }
      }
    }
  },
  {
    "_id": 3,
    "item": "Shampoo",
    "price": 5,
    "quantity": 10,
    "date": "2014-03-15T09:00:00Z",
    "props": {
      "brand": "Sunsilk",
      "size": "L",
      "seller": {
        "name": "Metro Chemicals Pvt. Ltd.",
        "address": {
          "city": "Mumbai",
          "pincode": 400004
        }
      }
    }
  },
  {
    "_id": 6,
    "item": "Comb",
    "price": 7.5,
    "quantity": 5,
    "date": "2015-06-04T05:08:13Z"
  },
  {
    "_id": 8,
    "item": "Soap",
    "price": 10,
    "quantity": 5,
    "date": "2016-02-06T20:20:13Z"
  },
  {
    "_id": 5,
    "item": "Soap",
    "price": 20,
    "quantity": 5,
    "date": "2014-04-04T21:23:13.331Z",
    "props": {
      "brand": "Lifebuoy",
      "size": "S",
      "seller": {
        "name": "Hans and Co.",
        "address": {
          "city": "Kolkata",
          "pincode": 700007
        }
      }
    }
  },
  {
    "_id": 1,
    "item": "Soap",
    "price": 10,
    "quantity": 2,
    "date": "2014-03-01T08:00:00Z",
    "props": {
      "brand": "Dettol",
      "size": "M",
      "seller": {
        "name": "Metro Chemicals Pvt. Ltd.",
        "address": {
          "city": "Mumbai",
          "pincode": 400004
        }
      }
    }
  },
  {
    "_id": 2,
    "item": "Mirror",
    "price": 20,
    "quantity": 1,
    "date": "2014-03-01T09:00:00Z"
  }
]
`
From this sorted list, we are applying pagination with offset 1 and limit 3. Hence, the result in "filter_with_sorting_and_pagination_response.json".
`Mirror` is last in the list so there is no way to get that since we have a limit 3.

The document in question
`{
    "_id": 4,
    "item": "Shampoo",
    "price": 5,
    "quantity": 20,
    "date": "2014-04-04T11:21:39.736Z"
  }` is not present since it is first in the sorted list and we use an offset of 1.

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java (hunk 6 above)
Comment (suresh-prakash) [reply document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:190]: Done. Will push the changes.

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java (hunk 7 above)
Comment (suresh-prakash) [reply document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:247]: Yes. Both are interchangeable. 
"setLimit()" and "setOffset()" are just helper methods to avoid additional (explicit) construction of the Pagination object.

Regarding having a single way, while that's possible,
- "setLimit()" and "setOffset()" are just additional utility methods, which internally build the Pagination object.
- Also, this (having 2 ways) is consistent with all other fields in the Query class. For example, "addSelection()" can be invoked either after constructing the SelectionSpec or by directly invoking with any SelectingExpression. 

In summary, every single member in the Query class has more than one way of defining for the purpose of "syntactic-ease" by avoiding redundant expansions. In other words, "setLimit()" and "setOffset()" are shorthand version of "setPagination()" similar to the other Query-class fields.

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java (hunk 8 above)
Comment (suresh-prakash) [reply document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java:260]: Yup. Just found that this is feasible in Postgres (Ref.: https://stackoverflow.com/a/7630564)

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/AggregationOperator.java (hunk 9 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/AggregationOperator.java:9]: Yup. Just found that this is feasible in Postgres (Ref.: https://stackoverflow.com/a/7630564)

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoSelectingExpressionParser.java (hunk 10 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoSelectingExpressionParser.java:65]: I would prefer to keep it "getProjectClause()" since this name (and all the other getXYZClause()) is based on the clause name in MongoDB. 

The clause in MongoDB for projection is "$project".

File: document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java (hunk 11 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java:70]: Yup. `filter` is similar to "WHERE" clause conditions, whereas `aggregationFilter` is similar to "HAVING" clause conditions.
For the Mongo specific case, `filter` is applied before "$group" and `aggregationFilter` is applied after "$group".
Ref.: MongoQueryExecutor:67

File: document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java (hunk 7 above)
Comment (kotharironak) [reply document-store/src/integrationTest/java/org/hypertrace/core/documentstore/mongo/MongoQueryExecutorTest.java]: As discussed, lets keep only as part of follow up PR.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoExpressionParser.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package org.hypertrace.core.documentstore.mongo.parser;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import org.hypertrace.core.documentstore.query.Query;
+
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
+public class MongoExpressionParser {
+  protected final Query query;
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoExpressionParser.java:9]: I have observed this it is not used. So, can we remove this as discussed (can be done in follow up)

**Review by kotharironak — APPROVED** [review]
